Machines API: CRDs and sample controller

pipejakob opened the pull request:
This PR has three commits:

1. An enormous update to our vendored packages, since I added more dependencies
2. Restructure the Machines API types to make them more useful
3. Add example CRD creation and example Machines controller

I demoed this all at today's Cluster API community breakout meeting, so if anyone missed it and wants to see these in action, the recording should be posted soon.

I want to vet that the first commit doesn't break anyone else's code before merging, but the other two should be safe to merge so we can start building on top of them or using them as examples. I'll sync with @ajitak and @jessicaochen to verify.

pipejakob committed:
Updating many vendor packages.

pipejakob committed:
Making Machines types more useful.

- Move Machines API types to api/machines/v1alpha1
- Add JSON annotations so that round-trip marshalling works
- Add helper code to create the Machines CRD (for the installer)
- Generate DeepCopy functions via deepcopy-gen

pipejakob committed:
Add example CRD creator and Machines controller.

These were demoed at today's Cluster API community breakout meeting.
There's a stand-alone example of how to add the Machines CRD to an
existing cluster, and a stand-alone controller with callback functions
to handle Machine create/update/delete events.

This is more to document progress and help sync with each other's
efforts than to actually start using now.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kube-deploy/309)
<!-- Reviewable:end -->

pipejakob committed:
Tiny patch to vendored kubicorn to fix build.

This is a short-term fix to unblock development. This patch should get
upstreamed properly, or we should fix our vendored dependencies.

pipejakob commented:
Okay, I added a tiny patch to the vendored version of kubicorn to unblock the build. I'll try to figure out if the proper fix is to upstream that patch to kubicorn or to rejigger our other vendored packages.
